### PR TITLE
feat(mission_planner): check goal footprint

### LINKED ISSUE #1761: behavior_path_planner generated path outside of drivable area
### Checklist

- [X] I've read the [contribution guidelines](https://github.com/autowarefoundation/autoware/blob/main/CONTRIBUTING.md).
- [X] I've searched other issues and no duplicate issues were found.
- [X] I'm convinced that this is not my fault but a bug.

### Description

If you give to EGO a goal pose a bit inclined , the path is still generated without considering of the width and length of EGO. And this generated path is outside the drivable area.

Screenshot from 2022-09-01 10-32-23

Screenshot from 2022-09-01 10-34-16

### Expected behavior

If the path to generated is outside the drivable area, the node should not create any path.

### Actual behavior

`behavior_path_planner` creates a path even if the path is outside the drivable area.

Screenshot from 2022-09-01 10-34-16

### Steps to reproduce

1. Download [1761.zip](https://github.com/autowarefoundation/autoware.universe/files/9469885/1761.zip)
2. Change pcd and lanelet2 maps file path in yaml file
3. run `ros2 launch scenario_test_runner scenario_test_runner.launch.py sensor_model:=sample_sensor_kit vehicle_model:=sample_vehicle \
scenario:=/scenario/path/1761.yaml \
architecture_type:=awf/universe launch_rviz:=false launch_autoware:=true`

4. Add ` /planning/scenario_planning/lane_driving/trajectory` topic in rviz

5. When you run the scenario, path occurred outside drivable area.

### Versions

_No response_

### Possible causes

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by ismetatabay**
Signed-off-by: ismet atabay <ismet@leodrive.ai>

## Description

I added checking function for if the goal's footprint is within the lane boundaries. Thus, if the footprint of the given goal extends beyond the lane boundaries, the goal will be rejected. Moreover, while the goal is published, the parameter is added for the maximum angle that the goal makes with the lane to be accepted. 

## Related links
closes #1761 
## Tests performed
Many goals were published on Rviz, it was seen that the goal was rejected when the goal's footprint was outside the lane boundaries. Goal footprint marker publisher is added for debugging purposes. In addition When the goal was given, it was tried with 20, 30, 45 degrees to see whether the goal would reject or not. Since the value is 45 when no parameter is added, 45 was entered as the default value. (I think 20°  better than 45°) 

https://user-images.githubusercontent.com/56237550/199095840-25d74459-d671-42e2-9352-a25e42dd0a78.mp4

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by ismetatabay** [comment]
I'm busy for the next 2 days, I will take care of it after thursday.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 1]
```diff
@@ -180,21 +184,60 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
+
+  const auto local_vehicle_footprint = createVehicleFootprint(vehicle_info_);
+  goal_footprint_ =
+    transformVector(local_vehicle_footprint, tier4_autoware_utils::pose2transform(goal));
+
+  // check if goal footprint exceeds lane
+  std::vector<tier4_autoware_utils::Point2d> points_intersection;
+  boost::geometry::intersection(
+    goal_footprint_, closest_lanelet.polygon2d().basicPolygon(), points_intersection);
+
```
[the remark this answers is not in the document]
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:233]: Hi, thanks for the suggestion, it made more sense to me as well. Now I have pushed the latest version.

**Comment by ismetatabay** [comment]
I think this PR is ready for review, thank you in advance. @takayuki5168 @isamu-takagi

File: planning/mission_planner/launch/mission_planner.launch.xml [hunk 2]
```diff
@@ -4,10 +4,12 @@
   <arg name="visualization_topic_name" default="/planning/mission_planning/route_marker"/>
 
   <node pkg="mission_planner" exec="mission_planner" name="mission_planner" output="screen">
+    <param from="$(find-pkg-share mission_planner)/config/mission_planner.param.yaml"/>
```
Comment (isamu-takagi) [inline planning/mission_planner/launch/mission_planner.launch.xml:7]: This file does not exist. I think you forgot to remove this line.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 3]
```diff
@@ -224,12 +275,11 @@ bool DefaultPlanner::is_goal_valid(const geometry_msgs::msg::Pose & goal) const
     const auto goal_yaw = tf2::getYaw(goal.orientation);
     const auto angle_diff = tier4_autoware_utils::normalizeRadian(lane_yaw - goal_yaw);
 
-    constexpr double th_angle = M_PI / 4;
+    const double th_angle = M_PI / (180 / p.goal_angle_threshold_deg);
```
Comment (isamu-takagi) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:278]: Using tier4_autoware_utils::deg2rad makes it more readable.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 4]
```diff
@@ -180,21 +189,63 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
   const auto goal_lanelet_pt = lanelet::utils::conversion::toLaneletPoint(goal.position);
 
   if (is_in_lane(closest_lanelet, goal_lanelet_pt)) {
     const auto lane_yaw = lanelet::utils::getLaneletAngle(closest_lanelet, goal.position);
     const auto goal_yaw = tf2::getYaw(goal.orientation);
     const auto angle_diff = tier4_autoware_utils::normalizeRadian(lane_yaw - goal_yaw);
 
-    constexpr double th_angle = M_PI / 4;
-
+    const double th_angle = M_PI / (180 / p.goal_angle_threshold_deg);
```
Comment (isamu-takagi) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:248]: Using `tier4_autoware_utils::deg2rad` makes it more readable.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.hpp [hunk 5]
```diff
@@ -51,11 +59,14 @@ class DefaultPlanner : public mission_planner::PlannerPlugin
   lanelet::ConstLanelets shoulder_lanelets_;
   route_handler::RouteHandler route_handler_;
 
+  DefaultPlannerParameters p;
```
Comment (isamu-takagi) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.hpp:62]: Could you rename it to something like `param_` to meet the [guidelines](https://autowarefoundation.github.io/autoware-documentation/main/contributing/coding-guidelines/languages/cpp/)?
- https://google.github.io/styleguide/cppguide.html#Variable_Names
- https://google.github.io/styleguide/cppguide.html#Variable_Comments

Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.hpp]: Thanks for the reviews, I implemented and pushed them all at latest commit.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 6]
```diff
@@ -180,21 +189,63 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
   }
+
+  const auto local_vehicle_footprint = createVehicleFootprint(vehicle_info_);
+  tier4_autoware_utils::LinearRing2d goal_footprint_ =
+    transformVector(local_vehicle_footprint, tier4_autoware_utils::pose2transform(goal));
+
+  // check if goal footprint exceeds lane
+  std::vector<tier4_autoware_utils::Point2d> points_intersection;
+  boost::geometry::intersection(
+    goal_footprint_, closest_lanelet.polygon2d().basicPolygon(), points_intersection);
```
Comment (isamu-takagi) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:232]: @takayuki5168 @TakaHoribe @mitsudome-r 
The closest may not be enough at the boundary of two lanelets. I'm not familiar with lanelet so I'd like to hear other comments.

Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:231]: I am planning to update that part as follows: Combining the next lanelets of the goal lanelet with the the previous lanes on the route. Then, i can check that polygon with goal footprint.

**Comment by ismetatabay** [comment]
When we think of edge scenarios, the footprint of the goal can be found on many lanelet. Also goal pose can be found in one lanelet, while the back or front of the vehicle boundaries can be found in another lanelet. To solve this case and another edge cases, I did it like this:   

- After the goal was published, I combined the created route lanelets (between goal pose and vehcile pose lanelets)
- Then, I combine them with goal' next lanelets (There are 3 different next lanes from the case 1 in the picture below, combining them separately)
- I check that the goal footprint in all combined lanelets.
- As a result, we accept the goal if the goal footprint is within the logical lane boundaries, otherwise we reject it.
I updated the description video for final results.

ss drawio

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 7]
```diff
@@ -92,6 +93,42 @@ bool is_in_parking_lot(
   return false;
 }
 
+bool is_footprint_in_lanelet(
+  const lanelet::ConstLanelet & goal_lanelet, tier4_autoware_utils::LinearRing2d goal_footprint,
+  lanelet::routing::RoutingGraphPtr routing_graph_ptr_, lanelet::ConstLanelets path_lanelets)
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:98]: add const and reference to all arguments?

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 8]
```diff
@@ -92,6 +93,42 @@ bool is_in_parking_lot(
...
+
+    auto combined_lanelets = combine_lanelets(lanelets);
+    boost::geometry::intersection(
+      goal_footprint, combined_lanelets.polygon2d().basicPolygon(), points_intersection);
+
+    if (points_intersection.empty()) {
+      return true;
+    }
+    points_intersection.clear();
+  }
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:127]: Here, only next lanelets are considered.
When lanelets longitudinal length is too short, not only the next lanelets but also the lanelet after the next, ... must be considered, right?

Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:127]: Thanks for your review, you are right, I forgot concerning about next lanelets of goal. I will think of a suitable approach for this situation.

**Comment by ismetatabay** [comment]
The footprint of the vehicle can be found in the next lanelets of the published goal lanelet. If these next lanelets are too short for the vehicle's length, the vehicle's footprint will be found in many lanelets (previous and next lanelets of goal lanelet).

It's uncomplicated to find previous lanelets where the vehicle's footprint is located, because the planned route includes these lanelets. I used a recursive search method to find the next lanelets. In this method, there is a search between the next lanes of the lane where the goal lane is located. The condition for stopping this search is that the length of the lanelets traveled exceeds the ```max_longitudinal_offset_m``` (wheel_base + front_overhang) of the vehicle. If the goal lanelet has more than one next lane, the search is repeated like this way.

To test the implementation, I divided the lanelet into small lanelets as described in the image below. I did not encounter any problems with the the algorithm.

split

**Comment by takayuki5168** [comment]
@isamu-takagi Could you review the architecture of mission_planner's implementation? I'll review if the check function works well .

**Review by isamu-takagi — COMMENTED** [review]
@takayuki5168 LGTM for architecture.

**Comment by mitsudome-r** [comment]
@takayuki5168 Were you able to check the function?

**Comment by takayuki5168** [comment]
Sorry to be late. I'll review this weekend.

**Comment by ismetatabay** [comment]
@takayuki5168 hi, The lanelet map that I made my own tests is as follows. There are quite a few little lanelets paired with linestring points on map.

[test_map.txt](https://github.com/autowarefoundation/autoware.universe/files/10092289/test_map.txt)


Selection_017

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 9]
```diff
@@ -14,15 +14,51 @@
...
 bool exists(const std::unordered_set<lanelet::Id> & set, const lanelet::Id & id)
 {
   return set.find(id) != set.end();
 }
 
+void appendPointToPolygon(
+  tier4_autoware_utils::Polygon2d & polygon, const tier4_autoware_utils::Point2d point)
+{
+  boost::geometry::append(polygon.outer(), point);
+}
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:33]: IMO: No need to create this function since this function has only one line.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 10]
```diff
@@ -255,15 +355,11 @@ PlannerPlugin::HADMapRoute DefaultPlanner::plan(const RoutePoints & points)
...
-    RCLCPP_WARN(logger, "Goal is not valid! Please check position and angle of goal_pose");
-    return route_msg;
-  }
-
+  lanelet::ConstLanelets path_lanelets;
   for (std::size_t i = 1; i < points.size(); i++) {
     const auto start_check_point = points.at(i - 1);
     const auto goal_check_point = points.at(i);
-    lanelet::ConstLanelets path_lanelets;
+    path_lanelets.clear();
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:362]: I feel this implementation had better be clean.

How about reverting this part of the change and give goal's lanelet to is_goal_valid.
Then is_goal_valid search previous lanelets if the footprint is inside the lanelets the same way as searching next lanelets.

Let me know if this idea does not sound good to you. I 'll explain more.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 11]
```diff
@@ -186,21 +194,114 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
+  tier4_autoware_utils::LinearRing2d goal_footprint =
+    transformVector(local_vehicle_footprint, tier4_autoware_utils::pose2transform(goal));
+  pub_goal_footprint_marker_->publish(visualize_debug_footprint(goal_footprint));
+  const auto polygon_footprint = convert_linear_ring_to_polygon(goal_footprint);
+
+  double next_lane_length = 0.0;
+  // combine calculated route lanelets
+  lanelet::ConstLanelet combined_prev_lanelet = combine_lanelets(path_lanelets);
+
+  // check if goal footprint exceeds lane and isn't in parking_lot
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:286]: > and isn't in parking_lot

I feel `when the goal is in parking_lot` makes us easier to understand the intention of this if condition. What do you think.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 12]
```diff
@@ -186,21 +194,114 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
+{
+  std::vector<tier4_autoware_utils::Point2d> points_intersection;
+
+  // check if goal footprint is in current lane
+  boost::geometry::intersection(
+    goal_footprint, combined_prev_lanelet.polygon2d().basicPolygon(), points_intersection);
+  if (points_intersection.empty()) {
+    return true;
+  }
+  points_intersection.clear();
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:236]: I think there is no need to clear here.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp [hunk 13]
```diff
@@ -186,21 +194,114 @@ PlannerPlugin::MarkerArray DefaultPlanner::visualize(const HADMapRoute & route)
...
+    if (vehicle_info_.max_longitudinal_offset_m + search_margin < next_lane_length) {
+      next_lane_length -= lanelet::utils::getLaneletLength2d(next_lane);
+      boost::geometry::intersection(
+        goal_footprint, combined_lanelets.polygon2d().basicPolygon(), points_intersection);
+      if (points_intersection.empty()) {
+        return true;
+      }
+      points_intersection.clear();
+    } else {  // if next lanelet length shorter than vehicle longitudinal offset -> recursive call
+      if (!check_goal_footprint(next_lane, combined_lanelets, goal_footprint, next_lane_length)) {
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: I feel that recursive call of a bit compress if/else condition is hard to debug, so I do not want to use it if we can.

In `check_goal_footprint`, how about implementing like
- 1. get goal_lanelet, and hold it as target_lanelets
- 2. add next lanelets from the goal to target_lanelets until the additional lane length is longer than vehicle_info_.max_longitudinal_offset_m + magin
- 3. add prev lanelets from the goal to target_lanelets until the additional lane length is longer than vehicle_info_.**min**_longitudinal_offset_m + magin
- 4. check if the goal footprint intersects with target_lanelets.

Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: First of all, many thanks for the review, I think it is necessary to write a recursive function in this point. Because when lanelets branches as in the case below, I think we need to use a recursive structure to search on all these cases.

Yes, you are right that the debugging part is difficult. To debug this function, I checked many things, such as the lanelets id's traversed on the edge cases, the length of the total traversed branches, etc. So in this final version I think everything is working as it should.

For such a search method, using path_lanelets as prev lanelets makes the algorithm simpler. If we also search for each prev_lanelets, we'll have to do a recursive search for each prev lanelet going to next lanelets

What do you think about this cases?

check_goal_footprint drawio

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 12 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:236]: I think it is necessary for some situations, for example if the vehicle is not inside the goal lanelet, points_intercestion will go to the search with non-empty, which means that even if there is no intersection in the new search, we will assume that there is an intersection because of these remaining points.

Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:236]: Before evaluating points_intersection on line 251, we overwrite points_intersection on line 249, 250.
Therefore I thought we do not need to clear points_intersection. Do I understand something wrong?
image

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 13 above)
Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: I understand the reason to use recursive call. Thank you very much for the good explanation.

---

Let me summarize the current collision checking (= if the path footprints are inside the lanelets or not) in autoware.universe.
With the current autoware.universe, the lanelets after the goal_lanelet are not considered for collision checking in 
- 1. mission_planner
- 2. obstacle_avoidance_planner
    - obstacle_avoidance_planner gets a drivable area from behavior_path_planner, and check if path footprints are inside the drivable area.
    - Drivable area is generated in behavior_path_planner using lanelets from the start to goal lanelet calculated in the mission_planner

With this PR, the mission planner will consider the lanelets after the goal for collision checking. So 1 will be fixed.

However, the behavior_path_planner, which generates the drivable area, does not know the correct lanelets after the goal in the case of your figure. Only mission_planner has the information of these lanelets.
Therefore, the drivable area can only have lanelets until the goal_lanelet, and the lanelets after the goal cannot be added to the drivable area. Therefore, 2 still exists.

IMO, in order to fix 2 (obstacle_avoidance_planner) as well as 1 (mission_planner), the goal lanelets should be the last lanelet in `check_goal_footprint` (maybe `next_lane` when returning true).
With this change, the drivable area can cover until the last lanelet after the goal considered in mission_planner, and 2 will be fixed.

In short, my suggestion is
- calculate the last lanelet recursively and hold it as `goal_lanelet`
    - This  implementation will be almost the same as `check_goal_footprint` in this PR.
- then, plan route between start lanelet to `goal_lanelet`.
- then, check if the goal footprint is inside the lanelets until `goal_lanelet`.


---

However, this change will be a bit large. I'm okay to merge the current PR without the further change to fix 1, and after that we will apply the change I suggested to fix 2.
Sorry to be too long. Feel free to share your opinion.

Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: (We can have a call if it's better for you.)

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 12 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:236]: @takayuki5168 hi, ```boost::geometry::intersection``` function doesn't overwrite points, it pushes back intersection points to existing ```point_intersection``` vector. Thus, if ```points_intersection``` go to the for loop section with non-empty, its falsely affects algorithm.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 13 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: First of all, sorry for the late conversion.

Thank you very much for your nice explanation, I understood the situation you said and we need to update it as well.

We can merge this PR if it's okay with you. I want to fix the 2nd situation by creating a new PR.

The methods you suggest can be done for this. 

- Additionally, I can create a new route by adding the traversed path up to the lane the goal is in to the path_lanelets.

- I was checking this path while debugging, it may be comfortable for me to implement it.

- As you said, I can create a new route between the last point where the goal lanelet is located and the point where the vehicle is located, that is also possible.

So how about I open a new PR with these? Or should I continue on this PR?

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 10 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:362]: If we don't want to use recursive call for previous lanes as we talked about before, it might be good to use path_lanelets for previous lanelets.

But in order to not affect the checkpoint mechanism, I made a small update here in the last commit.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 12 above)
Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:236]: I got it. Thanks for your explanation.

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 13 above)
Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: For me, it's okay to implement it on a new PR.
Then it seems that this PR is ready to merge?

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp [hunk 14]
```diff
@@ -14,15 +14,44 @@
...
+  const auto & i = vehicle_info;
 
+  LinearRing2d footprint;
+  footprint.push_back(Point2d{i.max_longitudinal_offset_m, i.max_lateral_offset_m});
+  footprint.push_back(Point2d{i.max_longitudinal_offset_m, i.min_lateral_offset_m});
+  footprint.push_back(Point2d{i.min_longitudinal_offset_m, i.min_lateral_offset_m});
+  footprint.push_back(Point2d{i.min_longitudinal_offset_m, i.max_lateral_offset_m});
+
+  return footprint;
+}
```
Comment (takayuki5168) [inline planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:43]: FYI: I will add this function to vehicle_info_util. With another PR, you can replace this function.
https://github.com/autowarefoundation/autoware.universe/pull/2394

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 13 above)
Comment (takayuki5168) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: Due to my late review, there happened a conflict.
I'm sorry, but could you resolve it. Then I will approve it.

File: planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp (hunk 14 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/utility_functions.cpp:43]: Thanks, i will update it :+1:

File: planning/mission_planner/src/lanelet2_plugins/default_planner.cpp (hunk 13 above)
Comment (ismetatabay) [reply planning/mission_planner/src/lanelet2_plugins/default_planner.cpp:256]: @takayuki5168 I solved conflicts, I thin it is ready to merge :+1:

**Comment by ismetatabay** [comment]
@takayuki5168 friendly ping :-)

**Comment by takayuki5168** [comment]
I will check if this PR works well tomorrow, and approve it.

**Comment by mitsudome-r** [comment]
@takayuki5168 Did you have chance to test the PR?

**Comment by takayuki5168** [comment]
@ismetatabay @mitsudome-r Sorry to be late. I confirmed that the PR works well.

**Comment by takayuki5168** [comment]
@ismetatabay I'm fine :)

**Comment by kosuke55** [comment]
we need to consider `shoulder lane`

**Comment by ismetatabay** [comment]
@kosuke55 Hi, I created PR: #2546

**Comment by kosuke55** [comment]
@ismetatabay 
thanks very much, I commented on your PR and I also send another PR https://github.com/autowarefoundation/autoware.universe/pull/2548
